File: web/src/api/assets.ts
import { requestJson, requestVoid } from './client'
import type { Asset, AssetContext, AssetType } from '../types/ops'

export type AssetPayload = {
  name: string
  asset_type: AssetType
  group_id?: number | null
  ssh_key_id?: number | null
  proxy_asset_id?: number | null
  host?: string
  port?: number
  username?: string
  auth_type?: string
  credential_secret?: string
  tags?: string[]
  vendor?: string
  description?: string
}

export type AssetConnectionTestResult = {
  success: boolean
  message: string
  detected_device_type: string | null
  detected_asset_type: string | null
  prompt: string | null
}

type AssetDto = {
  id: number
  group_id: number | null
  ssh_key_id: number | null
  proxy_asset_id: number | null
  name: string
  asset_type: AssetType
  host: string
  port: number
  username: string
  auth_type: string
  tags: string[]
  vendor: string
  description: string
}

type AssetContextDto = {
  asset: AssetDto
  terminal_session: {
    id: number
    status: string
    last_error: string
    started_at: string
    ended_at: string | null
  } | null
  recent_terminal_events: Array<{
    id: number
    event_type: string
    event_data: string
    created_at: string
  }>
  assistant_sessions: Array<{
    id: number
    title: string
    active_model: string
  }>
}

export function mapAsset(dto: AssetDto): Asset {
  return {
    id: dto.id,
    groupId: dto.group_id,
    sshKeyId: dto.ssh_key_id,
    proxyAssetId: dto.proxy_asset_id,
    name: dto.name,
    assetType: dto.asset_type,
    host: dto.host,
    port: dto.port,
    username: dto.username,
    authType: dto.auth_type,
    tags: dto.tags,
    vendor: dto.vendor,
    description: dto.description,
  }
}

export async function getAssets(): Promise<Asset[]> {
  const assets = await requestJson<AssetDto[]>('/api/assets')
  return assets.map(mapAsset)
}

export function testAssetConnection(asset: AssetPayload, assetId?: number | null): Promise<AssetConnectionTestResult> {
  return requestJson<AssetConnectionTestResult>('/api/assets/connection-test', {
    method: 'POST',
    body: JSON.stringify({ asset_id: assetId ?? null, asset }),
  })
}

export async function getAssetContext(assetId: number): Promise<AssetContext> {
  const context = await requestJson<AssetContextDto>(`/api/assets/${assetId}/context`)
  return {
    asset: mapAsset(context.asset),
    terminalSession: context.terminal_session
      ? {
          id: context.terminal_session.id,
          status: context.terminal_session.status,
          lastError: context.terminal_session.last_error,
          startedAt: context.terminal_session.started_at,
          endedAt: context.terminal_session.ended_at,
        }
      : null,
    recentTerminalEvents: context.recent_terminal_events.map((item) => ({
      id: item.id,
      eventType: item.event_type,
      eventData: item.event_data,
      createdAt: item.created_at,
    })),
    assistantSessions: context.assistant_sessions.map((item) => ({ id: item.id, title: item.title, model: item.active_model })),
  }
}

export async function createAsset(payload: AssetPayload): Promise<Asset> {
  const asset = await requestJson<AssetDto>('/api/assets', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
  return mapAsset(asset)
}

export async function updateAsset(assetId: number, payload: AssetPayload): Promise<Asset> {
  const asset = await requestJson<AssetDto>(`/api/assets/${assetId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
  return mapAsset(asset)
}

export async function deleteAsset(assetId: number): Promise<void> {
  await requestVoid(`/api/assets/${assetId}`, {
    method: 'DELETE',
  })
}

File: src/app/api/assets.py
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlmodel import Session

from app.api.schemas import AssetConnectionTestRequest, AssetConnectionTestView, AssetView
from app.db.session import get_session
from app.services.asset_service import (
    GroupNotFoundError,
    ProxyAssetInUseError,
    ProxyAssetInvalidError,
    ProxyAssetNotFoundError,
    SSHKeyNotFoundError,
    create_asset_record,
    delete_asset_record,
    get_asset_record,
    list_asset_records,
    update_asset_record,
)
from app.services.asset_connection_service import AssetConnectionService
from app.shared.schemas import AssetCreate

router = APIRouter()
_asset_connection_service = AssetConnectionService()


def to_asset_view(asset) -> AssetView:
    return AssetView(
        id=asset.id or 0,
        group_id=asset.group_id,
        ssh_key_id=asset.ssh_key_id,
        proxy_asset_id=asset.proxy_asset_id,
        name=asset.name,
        asset_type=asset.asset_type,
        host=asset.host,
        port=asset.port,
        username=asset.username,
        auth_type=asset.auth_type,
        tags=asset.tags.split(",") if asset.tags else [],
        vendor=asset.vendor,
        description=asset.description,
    )


@router.get("/api/assets")
def list_assets(session: Session = Depends(get_session)) -> list[AssetView]:
    return [to_asset_view(asset) for asset in list_asset_records(session)]


@router.post("/api/assets/connection-test", response_model=AssetConnectionTestView)
def test_asset_connection(
    payload: AssetConnectionTestRequest,
    session: Session = Depends(get_session),
) -> AssetConnectionTestView:
    try:
        result = _asset_connection_service.test(
            session,
            asset_id=payload.asset_id,
            asset_data=payload.asset,
        )
    except Exception as exc:
        return AssetConnectionTestView(success=False, message=str(exc))
    return AssetConnectionTestView.model_validate(result)


@router.get("/api/assets/{asset_id}")
def get_asset(asset_id: int, session: Session = Depends(get_session)) -> AssetView:
    asset = get_asset_record(session, asset_id)
    if asset is None:
        raise HTTPException(status_code=404, detail="Asset not found")
    return to_asset_view(asset)


@router.post("/api/assets", status_code=201)
def create_asset(payload: AssetCreate, session: Session = Depends(get_session)) -> AssetView:
    try:
        asset = create_asset_record(session, payload)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Group not found") from exc
    except SSHKeyNotFoundError as exc:
        raise HTTPException(status_code=404, detail="SSH key not found") from exc
    except ProxyAssetNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Proxy asset not found") from exc
    except ProxyAssetInvalidError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return to_asset_view(asset)


@router.put("/api/assets/{asset_id}")
def update_asset(asset_id: int, payload: AssetCreate, session: Session = Depends(get_session)) -> AssetView:
    try:
        asset = update_asset_record(session, asset_id, payload)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Group not found") from exc
    except SSHKeyNotFoundError as exc:
        raise HTTPException(status_code=404, detail="SSH key not found") from exc
    except ProxyAssetNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Proxy asset not found") from exc
    except ProxyAssetInvalidError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if asset is None:
        raise HTTPException(status_code=404, detail="Asset not found")
    return to_asset_view(asset)


@router.delete("/api/assets/{asset_id}", status_code=204)
def delete_asset(asset_id: int, session: Session = Depends(get_session)) -> Response:
    try:
        deleted = delete_asset_record(session, asset_id)
    except ProxyAssetInUseError as exc:
        raise HTTPException(status_code=409, detail=f"Proxy asset is used by {exc.dependent_count} asset(s)") from exc
    if not deleted:
        raise HTTPException(status_code=404, detail="Asset not found")
    return Response(status_code=204)
